zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -611,48 +580,67 @@ private[consumer] final class Runloop(
       _ => command.succeed.as(state.copy(subscription = command.subscription))
     )
 
-  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Unit]] =
-    ZStream
-      .mergeAll(3, 1)(
-        ZStream.repeatWithSchedule(Command.Poll, Schedule.fixed(pollFrequency)),
-        ZStream.fromQueue(requestQueue).mapChunks(c => Chunk.single(Command.Requests(c))),
-        ZStream.fromQueue(commandQueue)
-      )
-      .tap(cmd => diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)))
-      .runFoldZIO(State.initial) { (state, cmd) =>
-        ZIO.ifZIO(isShutdown)(onTrue = handleShutdown(state, cmd), onFalse = handleOperational(state, cmd))
-      }
+  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Any]] = {
+    def processCommands(state: State, wait: Boolean): Task[State] = for {
+      commands <- if (wait)
+                    commandQueue.takeBetween(1, commandQueueSize).timeoutTo(Chunk.empty)(identity)(pollFrequency)
+                  else commandQueue.takeAll // Gather available commands or return immediately if nothing in the queue
+
+      isShutdown <- isShutdown
+      updatedState <- ZIO.foldLeft(commands)(state) { case (s, cmd) =>
+                        (if (isShutdown) handleShutdown(s, cmd) else handleOperational(s, cmd)) <*
+                          diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))
+                      }
+    } yield updatedState
+
+    def doPollIfPendingActions(state: State): Task[(State, Boolean)] = {
+      val shouldPoll =
+        state.isSubscribed && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || state.assignedStreams.isEmpty)
+
+      for {
+        _ <-
+          ZIO
+            .logTrace(
+              s"Starting poll with ${state.pendingRequests.size} pending requests and ${state.pendingCommits.size} pending commits"
+            )
+            .when(shouldPoll)
+        newState <-
+          if (shouldPoll) handlePoll(state) else ZIO.succeed(state)
+      } yield (newState, !shouldPoll)
```

Comment:
What do you think of:
```suggestion
    def doPollIfPendingActions(state: State): Task[(State, Boolean)] = {
      def logPollStart: UIO[Unit] =
        ZIO
          .logTrace(
            s"Starting poll with ${state.pendingRequests.size} pending requests and ${state.pendingCommits.size} pending commits"
          )

      val shouldPoll =
        state.isSubscribed && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || state.assignedStreams.isEmpty)
      val wait = !shouldPoll
      if (shouldPoll) logPollStart *> handlePoll(state).map(_ -> wait) else ZIO.succeed(state -> wait)
    }
```